### PR TITLE
[CS] Allow ExprPatterns to be type-checked in the solver

### LINKED ISSUE #61850: `heterogeneous collection literal could only be inferred` warning for a homogeneous collection literal
Consider the following code:
```swift
class Base { }
class Derived : Base { }

let arr: [Base]

for case _ as [Derived] in [arr] {} // warning: heterogeneous collection literal could only be inferred to '[[Base]]'; add explicit type annotation if this is intentional
```

This diagnostics should not be emitted.

### LINKED ISSUE #63374: Move expression pattern solving into the constraint system
As discovered by the stress tester in https://github.com/apple/swift/pull/63022, we currently rely on CSApply calling `coercePatternToType` in order to type-check an expression pattern of a switch that's being solved in e.g a conjunction. This hampers solver-based code completion, preventing us from offering completions in expression pattern position. We ought to solve the expression as part of the constraint system.

rdar://104954155

### PR DESCRIPTION
**Pull request by hamishknight**
Previously we would wait until CSApply, which would trigger their type-checking in `coercePatternToType`. This caused a number of bugs, and hampered solver-based completion, which does not run CSApply. Instead, type-check ExprPatterns as a part of solving the overall pattern. We can then modify `coercePatternToType` to accept a closure, which allows the solver to take over rewriting the ExprPatterns it has already solved.

This then sets the stage for the complete removal of `coercePatternToType`, and doing all pattern type-checking in the solver.

This PR also includes a couple of cleanups/fixes for `getTypeForPattern`, and some other pattern crasher fixes.

Resolves #61850
Resolves #63374
rdar://104954155
rdar://105782480
rdar://106598067
rdar://107709341
rdar://107420031
rdar://109419240

### CODE REVIEW

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please test source compatibility

File: test/Constraints/patterns.swift [hunk 1]
```diff
@@ -564,3 +564,84 @@ struct TestIUOMatchOp {
...
+
+struct TestRecursiveVarRef<T> {
+  lazy var e: () -> Int = {e}()
+}
+
+func testMultiStmtClosureExprPattern(_ x: Int) {
+  if case { (); return x }() = x {}
+}
+
+func testExprPatternIsolation() {
```
Comment (hamishknight) [inline test/Constraints/patterns.swift:576]: @xedin Mind double checking that these cases make sense?

**Comment by hamishknight** [comment]
@swift-ci please test

File: lib/AST/ASTContext.cpp [hunk 2]
```diff
@@ -4178,7 +4178,6 @@ GenericFunctionType *GenericFunctionType::get(GenericSignature sig,
                                               Type result,
                                               Optional<ExtInfo> info) {
   assert(sig && "no generic signature for generic function type?!");
-  assert(!result->hasTypeVariable());
```
[the remark this answers is not in the document]
Comment (slavapestov) [reply lib/AST/ASTContext.cpp:4181]: As long as we allocate the GenericFunctionType in the correct arena it is probably fine. However we cannot allow type variables in the GenericSignature for more fundamental reasons.

File: test/Constraints/patterns.swift [hunk 3]
```diff
@@ -564,3 +564,84 @@ struct TestIUOMatchOp {
...
+  let _ = {
+    if case 0 = nil {} // expected-error {{'nil' requires a contextual type}}
+  }
+  for case 0 in nil {} // expected-error {{'nil' requires a contextual type}}
+  for case 0 in [nil] {}
+  // expected-error@-1 {{type 'Any' cannot conform to 'Equatable'}}
+  // expected-note@-2 {{only concrete types such as structs, enums and classes can conform to protocols}}
+  // expected-note@-3 {{requirement from conditional conformance of 'Any?' to 'Equatable'}}
+
+  // Though we will try Double for an integer literal...
```
Comment (xedin) [inline test/Constraints/patterns.swift:588]: Is this specific to `Double` or `Int` works as well (I expect it does)?

**Review by xedin — APPROVED** [review]
LGTM! It might make sense to have an alternative version of this where we either use a TypeJoinExpr to type-check expr patterns that convert to the same generic parameter or simply enable bi-directional checking for patterns since each pattern is solved separately anyway and it doesn't seem like they are too complex.

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please test source compatibility

**Comment by hamishknight** [comment]
@swift-ci please test

**Review by xedin — APPROVED** [review]

File: lib/Sema/TypeCheckPattern.cpp [hunk 4]
```diff
@@ -1034,15 +1034,61 @@ void repairTupleOrAssociatedValuePatternIfApplicable(
...
+      auto *NoneEnumElement = ctx.getOptionalNoneDecl();
+      auto *BaseTE = TypeExpr::createImplicit(patternTy, ctx);
+      auto *EEP = new (ctx)
+          EnumElementPattern(BaseTE, NLE->getLoc(), DeclNameLoc(NLE->getLoc()),
+                             NoneEnumElement->createNameRef(), NoneEnumElement,
+                             nullptr, EP->getDeclContext());
+      EEP->setType(patternTy);
+      return EEP;
+    } else {
+      // ...but for non-optional types it can never match! Diagnose it.
```
[the remark this answers is not in the document]
Comment (hamishknight) [reply lib/Sema/TypeCheckPattern.cpp:1065]: Yes, but we also need to diagnose for cases that don't yet go through the constraint system. We should be able to move this into the constraint system once we delete `coercePatternToType`

**Comment by hamishknight** [comment]
Source compat failures match those on main

File: lib/Sema/TypeCheckPattern.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (xedin) [reply lib/Sema/TypeCheckPattern.cpp:1065]: Fair enough! We could start with cases that do go through the solver and expand later too :)

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please test source compatibility

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please test source compatibility

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please test source compatibility

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please test source compatibility

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please test Windows

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please SourceKit stress test

**Comment by hamishknight** [comment]
@xedin Mind taking a look at 406354935e93ebff5a9c08b7b9055e3560ee77e6? I _believe_ that should be the only change since you last looked at this

**Review by xedin — APPROVED** [review]

**Comment by hamishknight** [comment]
@swift-ci please smoke test and merge

**Comment by hamishknight** [comment]
Going to stick to the conjunction implementation for now

**Comment by hamishknight** [comment]
@swift-ci please smoke test and merge

**Comment by hamishknight** [comment]
@swift-ci please smoke test and merge

**Pull request merged**